src/jit/gentree.cpp

```diff
@@ -15432,10 +15450,7 @@ bool GenTree::DefinesLocal(Compiler* comp, GenTreeLclVarCommon** pLclVarTree, bo
             blkNode = gtOp.gtOp1->AsBlk();
         }
     }
-    else if (OperIsBlk())
```

Comment:
This is here to catch the block store case, but obviously needs to only check for stores. I'm not sure how this can be correct if it's only checking for assignments, and this is used in liveness.

Reply:
Looks at first glance like many callers use `DefinesLocal` to try and determine if we're going to track byref exposed separately from gc heap exposed. So returning false more often just makes the jit more conservative.

For correctness, Liveness uses `DefinesLocalAddr` and looks for `GTF_VAR_DEF` to determine context.

So I suppose we could update morph to use `DefinesLocalAddr` and flags checks to kill assertions, and separately determine what if anything we need to do to `DefinesLocal`.

Reply:
Or use `DefinesLocal` and likewise do a flag check, which is what I more or less did initially.

Reply:
Looks to me that instead of `OperIsBlk` we need to specifically check for GT_STORE_OBJ/BLK/DYN_BLK

Reply:
Right, that's what I meant to say above, but obviously wasn't clear.

Reply:
This just shows that better handling of assignment to struct typed fields is a good preamble to getting rid of `GT_ASG`. Once that's done we only need to check for `GT_ASG(GT_LCL_VAR/FLD, x)` and `GT_STORE_LCL_VAR/FLD`. Without the whole `DefinesLocalAddr` goo.

Reply:
I'm starting to think that the best fix at the moment is to wrap the `DefinesLocal` call in morph in `if (tree->OperIs(GT_ASG))`. All its over uses are under similar guards.

Even if you make the `OperIsBlk` check more specific, this function still doesn't cover all "defines local" cases - you'll need to add `GT_STORE_LCL_VAR/FLD` to it for what. And maybe `GT_STOREIND` too.

Ultimately this function is an artifact of `GT_ASG` and it should disappear together with it.

Reply:
I'd be fine with that too. Gives the same diffs as the other two variants I've tried.